File: include/thumbnail/Thumbnail.php
<?php

namespace thumbnail;

use asset\Asset;
use database\Database;
use log\Log;

use GdImage;
use log\LogLevel;
use RuntimeException;
use SQLite3Result;
use thumbnail\ThumbnailFormat;

class Thumbnail
{

	private static function getThumbnailStorePath(): string
	{
		return __DIR__ . '/../../public/thumbnail';
	}

	/**
	 * Deletes all thumbnail variations carrying asset ids no longer in the database.
	 * @return void 
	 */
	public static function deleteOrphanedThumbnails(): void
	{
		// Do nothing if the thumbnail directory does not exist
		if (!is_dir(self::getThumbnailStorePath())) {
			return;
		}

		// Get all existing asset IDs from the database
		$existingIds = [];
		$dbResult = Database::runQuery("SELECT id FROM Asset");
		assert($dbResult instanceof SQLite3Result);
		while ($row = $dbResult->fetchArray()) {
			$existingIds[] = $row['id'];
		}


		$thumbnailDir = self::getThumbnailStorePath() . "/";
		foreach (scandir($thumbnailDir) as $variationDir) {
			if ($variationDir === '.' || $variationDir === '..') {
				continue;
			}
			$fullVariationDir = $thumbnailDir . $variationDir . "/";
			foreach (scandir($fullVariationDir) as $file) {
				if ($file === '.' || $file === '..') {
					continue;
				}
				$assetId = intval(pathinfo($file, PATHINFO_FILENAME));
				if (!in_array($assetId, $existingIds)) {
					unlink($fullVariationDir . $file);
					Log::write("Deleted orphaned thumbnail", $fullVariationDir . $file, LogLevel::DEBUG);
				}
			}
		}
	}

	public static function saveThumbnailVariations(int $assetId, GdImage $originalImage): void
	{

		// Is the input image valid?
		self::validateImage($originalImage);

		foreach (ThumbnailFormat::cases() as $t) {

			$gdImage = Thumbnail::createThumbnailFromImageData($originalImage, $t);

			$fileName = self::getThumbnailStorePath() . "/" .
				$t->value . "/$assetId." . strtolower($t->getExtension());

			// Create directory if it does not exist
			$directory = dirname($fileName);
			if (!is_dir($directory)) {
				mkdir($directory, 0755, true);
			}

			// Save image
			match ($t->getExtension()) {
				"JPG" => imagejpeg($gdImage, $fileName, 95),
				"PNG" => imagepng($gdImage, $fileName, 6),
				default => throw new \InvalidArgumentException("Unsupported image format: " . $t->getExtension()),
			};

			self::validateImageFile($fileName);

			Log::write("Saved thumbnail", ["assetId" => $assetId, "fileName" => $fileName], LogLevel::DEBUG);
		}
	}

	private static function validateImageFile(string $filePath): void
	{
		// Exists?
		if (!file_exists($filePath)) {
			throw new RuntimeException("Thumbnail file does not exist: " . $filePath);
		}

		// Not empty?
		if (filesize($filePath) === 0) {
			throw new RuntimeException("Thumbnail file is empty: " . $filePath);
		}

		$imageInfo = getimagesize($filePath);
		if ($imageInfo === false) {
			throw new RuntimeException("Failed to get image info for thumbnail: " . $filePath);
		}

		$gdImage = match ($imageInfo[2]) {
			IMAGETYPE_JPEG => imagecreatefromjpeg($filePath),
			IMAGETYPE_PNG => imagecreatefrompng($filePath),
			default => throw new RuntimeException("Unsupported thumbnail image type: " . $filePath),
		};

		if ($gdImage === false) {
			throw new RuntimeException("Failed to create GD image for validation purposes: " . $filePath);
		}

		self::validateImage($gdImage);
	}

	private static function validateImage(GdImage $gdImage): void
	{

		// Not identical on all pixels?
		// I.e. is the entire image black/white/transparent?

		$width = imagesx($gdImage);
		$height = imagesy($gdImage);

		if ($width === 0 || $height === 0) {
			throw new RuntimeException("Thumbnail image has zero width or height, likely invalid.");
		}

		$checkInterval = max(1, min((int)($width / 10), (int)($height / 10)));
		$allPixelsSame = true;
		$firstPixel = imagecolorat($gdImage, 0, 0);
		for ($x = 0; $x < $width; $x += $checkInterval) {
			for ($y = 0; $y < $height; $y += $checkInterval) {
				if (imagecolorat($gdImage, $x, $y) !== $firstPixel) {
					$allPixelsSame = false;
					break 2;
				}
			}
		}

		if ($allPixelsSame) {
			throw new RuntimeException("Thumbnail image is uniformly colored and likely invalid");
		}
	}

	public static function createThumbnailFromImageData(GdImage $rawImage, ThumbnailFormat $format): GdImage
	{

		Log::write("Building variation " . $format->value, LogLevel::DEBUG);

		$originalWidth = imagesx($rawImage);
		$originalHeight = imagesy($rawImage);

		// Calculate new dimensions maintaining aspect ratio
		$ratio = min($format->getSize() / $originalWidth, $format->getSize() / $originalHeight);
		$newWidth = (int)($originalWidth * $ratio);
		$newHeight = (int)($originalHeight * $ratio);

		// Calculate offsets to center the image
		$offsetX = (int)(($format->getSize() - $newWidth) / 2);
		$offsetY = (int)(($format->getSize() - $newHeight) / 2);

		// Create output image
		$outputImage = imagecreatetruecolor($format->getSize(), $format->getSize());

		if ($format->getBackgroundColorHex() !== NULL) {

			// Fill with background color
			$r = max(min(intval(substr($format->getBackgroundColorHex(), 0, 2), 16), 255), 0);
			$g = max(min(intval(substr($format->getBackgroundColorHex(), 2, 2), 16), 255), 0);
			$b = max(min(intval(substr($format->getBackgroundColorHex(), 4, 2), 16), 255), 0);
			$bgColor = imagecolorallocate($outputImage, $r, $g, $b);
			if ($bgColor === false) {
				throw new RuntimeException("Failed to allocate background color.");
			}
			imagefill($outputImage, 0, 0, $bgColor);
		} else {

			// Transparent background
			imagealphablending($outputImage, false);
			imagesavealpha($outputImage, true);
			$transparent = imagecolorallocatealpha($outputImage, 0, 0, 0, 127);
			if ($transparent === false) {
				throw new RuntimeException("Failed to allocate transparent color.");
			}
			imagefill($outputImage, 0, 0, $transparent);
			imagealphablending($outputImage, true);
		}

		// Resize and copy the original image centered
		imagealphablending($rawImage, true);
		imagecopyresampled($outputImage, $rawImage, $offsetX, $offsetY, 0, 0, $newWidth, $newHeight, $originalWidth, $originalHeight);

		return $outputImage;
	}
}

File: app/thumbnail/thumbnail.py
from pathlib import Path
from typing import Optional

from PIL import Image, ImageDraw

from app.log.log import Log
from app.log.log_level import LogLevel
from app.thumbnail.thumbnail_format import ThumbnailFormat


class Thumbnail:
    @staticmethod
    def _get_store_path() -> Path:
        return Path(__file__).parent.parent.parent / "public" / "thumbnail"

    @classmethod
    def delete_orphaned_thumbnails(cls) -> None:
        from app.database.database import Database

        store_path = cls._get_store_path()
        if not store_path.is_dir():
            return

        result = Database.run_query("SELECT id FROM Asset")
        existing_ids = {row[0] for row in result.fetchall()}

        for variation_dir in store_path.iterdir():
            if not variation_dir.is_dir():
                continue
            for file in variation_dir.iterdir():
                try:
                    asset_id = int(file.stem)
                    if asset_id not in existing_ids:
                        file.unlink()
                        Log.write("Deleted orphaned thumbnail", str(file), LogLevel.DEBUG)
                except (ValueError, OSError):
                    pass

    @classmethod
    def save_thumbnail_variations(cls, asset_id: int, original_image: Image.Image) -> None:
        cls._validate_image(original_image)

        for fmt in ThumbnailFormat:
            thumb = cls.create_thumbnail_from_image(original_image, fmt)
            ext = fmt.get_extension().lower()
            file_path = cls._get_store_path() / fmt.value / f"{asset_id}.{ext}"
            file_path.parent.mkdir(parents=True, exist_ok=True)

            if fmt.get_extension() == "JPG":
                # Convert RGBA to RGB with background fill before saving as JPEG
                rgb_img = Image.new("RGB", thumb.size, (255, 255, 255))
                if thumb.mode == "RGBA":
                    rgb_img.paste(thumb, mask=thumb.split()[3])
                else:
                    rgb_img.paste(thumb)
                rgb_img.save(str(file_path), "JPEG", quality=95)
            else:
                thumb.save(str(file_path), "PNG", compress_level=6)

            cls._validate_image_file(file_path)
            Log.write("Saved thumbnail", {"asset_id": asset_id, "file": str(file_path)}, LogLevel.DEBUG)

    @classmethod
    def _validate_image_file(cls, file_path: Path) -> None:
        if not file_path.exists():
            raise RuntimeError(f"Thumbnail file does not exist: {file_path}")
        if file_path.stat().st_size == 0:
            raise RuntimeError(f"Thumbnail file is empty: {file_path}")
        try:
            img = Image.open(str(file_path))
            img.verify()
        except Exception as e:
            raise RuntimeError(f"Thumbnail file is invalid: {file_path}: {e}") from e
        # Re-open after verify (verify closes the image)
        img = Image.open(str(file_path)).convert("RGBA")
        cls._validate_image(img)

    @staticmethod
    def _validate_image(image: Image.Image) -> None:
        img = image.convert("RGBA")
        width, height = img.size
        check_interval = 4
        first_pixel = img.getpixel((0, 0))
        all_same = True
        for x in range(0, width, check_interval):
            for y in range(0, height, check_interval):
                if img.getpixel((x, y)) != first_pixel:
                    all_same = False
                    break
            if not all_same:
                break
        if all_same:
            raise RuntimeError("Thumbnail image is uniformly colored and likely invalid")

    @staticmethod
    def create_thumbnail_from_image(raw_image: Image.Image, fmt: ThumbnailFormat) -> Image.Image:
        size = fmt.get_size()
        orig_w, orig_h = raw_image.size

        ratio = min(size / orig_w, size / orig_h)
        new_w = int(orig_w * ratio)
        new_h = int(orig_h * ratio)

        offset_x = (size - new_w) // 2
        offset_y = (size - new_h) // 2

        bg_hex = fmt.get_background_color_hex()
        if bg_hex:
            r = int(bg_hex[0:2], 16)
            g = int(bg_hex[2:4], 16)
            b = int(bg_hex[4:6], 16)
            output = Image.new("RGB", (size, size), (r, g, b))
        else:
            output = Image.new("RGBA", (size, size), (0, 0, 0, 0))

        resized = raw_image.resize((new_w, new_h), Image.LANCZOS)
        if bg_hex:
            if resized.mode == "RGBA":
                output.paste(resized.convert("RGB"), (offset_x, offset_y))
            else:
                output.paste(resized, (offset_x, offset_y))
        else:
            if resized.mode != "RGBA":
                resized = resized.convert("RGBA")
            output.paste(resized, (offset_x, offset_y), resized)

        return output
